hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractMerger.java

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.impl;
+
+import com.hazelcast.config.InMemoryFormat;
+import com.hazelcast.core.ExecutionCallback;
+import com.hazelcast.internal.cluster.ClusterService;
+import com.hazelcast.internal.cluster.Versions;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.nio.Address;
+import com.hazelcast.nio.Disposable;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.OperationFactory;
+import com.hazelcast.spi.OperationService;
+import com.hazelcast.spi.SplitBrainHandlerService;
+import com.hazelcast.spi.SplitBrainMergePolicy;
+import com.hazelcast.spi.merge.MergingEntryHolder;
+import com.hazelcast.spi.partition.IPartitionService;
+import com.hazelcast.util.MutableLong;
+import com.hazelcast.util.function.BiConsumer;
+import com.hazelcast.version.Version;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+import static com.hazelcast.config.InMemoryFormat.NATIVE;
+import static com.hazelcast.util.ExceptionUtil.rethrow;
+import static java.lang.String.format;
+
+/**
+ * Used by both icache and imap to provide a shared merge runnable for {@link SplitBrainHandlerService#prepareMergeRunnable()}
+ *
+ * @param <S> type of the store in a partition
+ */
+public abstract class AbstractMerger<S> implements Runnable, Disposable {
```

Comment:
I'm thinking if `AbstractEntryMerger` would be a better name, since this one is specialized for key-based data structures with a `MergingEntryHolder`. Maybe we can also create an `AbstractValueMerge` in another PR for the key-less data structures.

Reply:
changed to use generics instead of `MergingEntryHolder`